### ISSUE TITLE
Questions about `Arbitrary`

### ISSUE THREAD

**Issue by davidbarsky** [rust-fuzz/arbitrary#27]
Hi! A few notes to preface issue:
- I asked @Manishearth this question a few weeks back and said that I should open an issue.
- This is _not_ meant to be judgement of the decisions made in this crate, especially since I am not aware of the constraints that y'all might have.

The question: why did y'all opt for a QuickCheck-style, trait-based approach over a Hypothesis/Proptest-style, explicit shrinking object/struct approach? For context, a Hypothesis/Proptest-style checker would have a shrinking strategy that returns a tree. This tree would values and shrinking/expansion strategies. I've noticed that the biggest win of Hypothesis/Proptest-style approaches is that shrinkers/generators can be implemented for completely arbitrary (no pun intended!) types so that orphan rules don't apply to generated code.

That being said, I suspect that fuzzers might have different constraints than property testing, so this entire question could be taken with a grain of salt.

If I recall correctly, @fitzgen mentioned to me that they preferred Quickcheck to Proptest-style systems at Strangeloop.

**Comment by Manishearth** [comment]
For a start, while we could indeed do this for _shrinking_ , this wouldn't be as useful for arbitrary generation, since here it needs to be pretty fast and ideally not too complicated.

Never thought super deeply about applying the proptest model here, though.

**Comment by nagisa** [comment]
> The question: why did y'all opt for a QuickCheck-style, trait-based approach

This crate originally was a half-assed extract of the trait with the same name from `quickcheck` crate, and then ability to generate data from a ringbuffer was added so that this could be used with a fuzzer.

Hence the similarities.

**Comment by fitzgen** [comment]
Let me preface this by saying that I am by no means an absolute expert here.

I think we could indeed get more effective shrinking with `proptest`-style traits than we have with `Arbitrary::shrink`. However, 95% of the time, we aren't shrinking the instance directly with `Arbitrary::shrink`, but instead asking libfuzzer to shrink its input string that we then generate instances from. I suspect this works as well in practice as the `proptest` method of shrinking. It has definitely been Good Enough for me, and I use `Arbitrary::shrink` as a backup method.

I don't think we would get more effective testcase generation with `proptest`, nor do we really care about enabling multiple different test case generation strategies like `proptest` does, because we have libfuzzer and its insight into code coverage to guide us.

So basically, the way I see it, if you're doing purely random generation-based testing, then `proptest` style traits are the way to go here (although you probably also want [swarm testing](https://www.cs.utah.edu/~regehr/papers/swarm12.pdf) in addition to that).

However, if you have a coverage-guided, mutation-based fuzzer that is guiding you, then all that extra stuff is unnecessary and you want test case generation to be as thin a veneer over the fuzzer's input string as possible. This is the direction we've been moving this crate in, with the recent `Unstructured` overhaul and all that.

**Comment by davidbarsky** [comment]
Thanks for explaining that, @fitzgen, @nagisa, and @Manishearth!

> I don't think we would get more effective testcase generation with proptest, nor do we really care about enabling multiple different test case generation strategies like proptest does, because we have libfuzzer and its insight into code coverage to guide us.

That makes sense. The usecase I thinking about—and I am _even less_ of an expert than what you claim—would be to reuse the same generators between property tests and a fuzzer (which would be a standalone target from Cargo's perspective). The property tests could be driven by a `rand::Rng` and the fuzzer targets could be driven by the `Unstructured` stuff.

> However, if you have a coverage-guided, mutation-based fuzzer that is guiding you, then all that extra stuff is unnecessary and you want test case generation to be as thin a veneer over the fuzzer's input string as possible. This is the direction we've been moving this crate in, with the recent Unstructured overhaul and all that.

Makes sense. If I'm understanding you correctly, the usecase I described might have a bit of an impedance mismatch which would be difficult to bridge.

**Comment by fitzgen** [comment]
Yeah, it sounds like you're doing something very similar to [what I was doing not too long ago](https://fitzgeraldnick.com/2019/09/04/combining-coverage-guided-and-generation-based-fuzzing.html), and my conclusions were that I could get the most bang for my buck by investing in this crate to make structure-aware testing really good with `arbitrary`, `libfuzzer`, and `cargo fuzz`.

Having a bridge is nice for porting targets, but I don't think that it is worth maintaining support for running on both quickcheck's or proptest's driver and on top of AFL or libfuzzer in the long term. I personally think that just focusing on the latter is the way to go.

**Comment by Manishearth** [comment]
Yeah so I've wanted to make it so that fuzz tests are quickcheck/proptest tests as well, it's just annoying to figure out how to set up.

**Comment by davidbarsky** [comment]
Makes sense! Thanks for all your answers :)
